File: c04/c04/Stack.cpp
#include "Stack.h"
/*
StackItem::StackItem(char key, StackItem* pNext) {
    setKey(key);
    setNext(pNext);
}

StackItem:: ~StackItem() {
 
} 

Stack::Stack() {
  this->m_pHead = nullptr;
}

Stack::~Stack() {
  while( !isEmpty()) {
    del();
  }
}

void Stack::push( char c ) {
  StackItem* p = new ( nothrow ) StackItem(c, m_pHead); // jesli blad = null
  if (!p) {
  throw StackException(STACK_UNDEF_ERROR);
  }

  m_pHead = p;
  }

char Stack::pop() {
  if (!isEmpty()) {
  char c = top();
  del();
  return c;
  }
  throw StackException(STACK_UNDERFLOW);
}

char Stack::top() {
  if (!isEmpty()) {
  return m_pHead->getKey();
  }
  throw StackException(STACK_UNDERFLOW);
}

void Stack::del() {
  if (!isEmpty()) {
  StackItem* p = this->m_pHead;
  m_pHead = p->getNext();
  delete p;
  }
  else
    throw StackException(STACK_UNDERFLOW);
}
*/

File: StackT/src/Stack.java

enum ErrCode { STACK_ALLOCATION(1), STACK_IS_EMPTY(2), OTHER_ERROR(0);
    private int mCode;
    ErrCode(int code) { mCode = code; }
}

class StackException extends RuntimeException //lub Exception
{
    private ErrCode mErrCode;
    StackException( ErrCode errCode) { mErrCode = errCode; }
    String getReason()
    {
        switch( mErrCode ) {
            case STACK_ALLOCATION:
                return "ERROR: Allocation error";
            case STACK_IS_EMPTY:
                return "ERROR: Stack is empty";
            default: return "Other error";
        }
    }
}

public class Stack<T> {

    private class StackItem<T>
    {
        private T mKey;
        private StackItem<T> mNext;
        public StackItem(T obj, StackItem next)
        { mKey = obj; mNext = next; }

        //-----gettery-i-settery-----
        public T getKey() {
            return mKey;
        }

        public final void setKey(T key) {
            this.mKey = key;
        }

        public StackItem<T> getNext() {
            return mNext;
        }

        public final void setNext(StackItem<T> next) {
            this.mNext = next;
        }
    }

    private StackItem<T> pHead = null;

    public Stack() {}

    public void push(T item) {
        try {
        pHead = new StackItem<T>(item,pHead); }
        catch ( OutOfMemoryError e ) {
            throw new StackException(ErrCode.STACK_ALLOCATION);
        }
    }

    public T top() {
        if (isEmpty()) {
            throw new StackException(ErrCode.STACK_IS_EMPTY);
        }
        return pHead.getKey();
    }

    public void del() {
        if (isEmpty()) {
            throw new StackException(ErrCode.STACK_IS_EMPTY);
        }
        pHead = pHead.getNext();
    }

    public T pop() {

        if (isEmpty()) {
            throw new StackException(ErrCode.STACK_IS_EMPTY);
        }

        T info = top();
        del();
        return info;
    }

    public boolean isEmpty() {
        return pHead == null;
    }
}
